File: matlab/stereo2_DP.m
% Stereo Matching using Dynamic Programming - smoothness cost function approach
% Computes a disparity map from a rectified stereo pair using Dynamic Programming

function stereo2_DP()

    % Set parameters
    dispLevels = 16; %disparity range: 0 to dispLevels-1
    lambda = 10; %weight of smoothness cost
    trunc = 2; %truncation of smoothness cost

    % Define matching cost function
    computeMatchingCost = @(left,right) abs(left-right); %absolute differences

    % Define smoothness cost function
    computeSmoothnessCost = @(d1,d2) lambda*min(abs(d1-d2),trunc);

    % Load left and right images in grayscale
    leftImg = rgb2gray(imread('left.png'));
    rightImg = rgb2gray(imread('right.png'));

    % Apply a Gaussian filter
    leftImg = imgaussfilt(leftImg,0.6,'FilterSize',5);
    rightImg = imgaussfilt(rightImg,0.6,'FilterSize',5);

    % Get the size
    [rows,cols] = size(leftImg);

    % Convert to int32
    leftImg = int32(leftImg);
    rightImg = int32(rightImg);

    % Compute pixel-based matching costs
    matchingCosts = zeros(rows,cols,dispLevels,'int32');
    for d = 0:dispLevels-1
        rightImgShifted = shiftRight(rightImg,d,0);
        matchingCosts(:,:,d+1) = computeMatchingCost(leftImg,rightImgShifted);
    end

    % Compute smoothness costs
    d = 0:dispLevels-1;
    smoothnessCosts = computeSmoothnessCost(d,d.');
    smoothnessCosts = permute(int32(smoothnessCosts),[3 4 1 2]);

    % Initialize minimum cost paths and transitions for the left to right direction
    fromLeft = zeros(rows,cols,dispLevels,'int32');
    transitions = zeros(rows,cols,dispLevels,'int32');

    % Compute minimum cost paths and transitions for left to right direction
    for x = 1:cols-1
        currentCosts = matchingCosts(:,x,:) + fromLeft(:,x,:);
        [C,T] = computeDirectionalCosts(currentCosts,smoothnessCosts);
        fromLeft(:,x+1,:) = C;
        transitions(:,x+1,:) = T;
    end

    % Compute the disparity map - Backtracking
    dispMap = zeros(rows,cols);
    [~,ind] = min(fromLeft(:,cols,:),[],3);
    for x = cols:-1:1
        dispMap(:,x) = ind-1;
        ind = transitions(sub2ind(size(transitions),(1:rows).',x*ones(rows,1),ind)); %get the disparity transitions
    end

    % Normalize the disparity map for display
    scaleFactor = 256/dispLevels;
    dispImg = uint8(dispMap*scaleFactor);

    % Show disparity map
    figure; imshow(dispImg)

    % Save disparity map
    imwrite(dispImg,'disparity2_DP.png')
end

% Compute minimum cost paths and transitions
% ------------------------------------------
function [output,transitions] = computeDirectionalCosts(currentCosts,smoothnessCosts)
    sum = currentCosts + smoothnessCosts;
    output = permute(min(sum,[],3),[1 2 4 3]);
    output = output - min(output,[],3); %normalize
    [~,ind] = min(sum,[],3);
    transitions = int32(permute(ind,[1 2 4 3]));
end

% Shift Functions (Down/Up/Right/Left/Forward/Backward)
% -----------------------------------------------------
function B = shiftDown(A,n,fillValue)
    B = circshift(A,n,1);
    B(1:min(n,end),:,:) = fillValue;
end
function B = shiftUp(A,n,fillValue)
    B = circshift(A,-n,1);
    B(max(end-n+1,1):end,:,:) = fillValue;
end
function B = shiftRight(A,n,fillValue)
    B = circshift(A,n,2);
    B(:,1:min(n,end),:) = fillValue;
end
function B = shiftLeft(A,n,fillValue)
    B = circshift(A,-n,2);
    B(:,max(end-n+1,1):end,:) = fillValue;
end
function B = shiftForward(A,n,fillValue)
    B = circshift(A,n,3);
    B(:,:,1:min(n,end)) = fillValue;
end
function B = shiftBackward(A,n,fillValue)
    B = circshift(A,-n,3);
    B(:,:,max(end-n+1,1):end) = fillValue;
end

File: python/stereo2_DP.py
# Stereo Matching using Dynamic Programming - smoothness cost function approach
# Computes a disparity map from a rectified stereo pair using Dynamic Programming

import numpy as np
import cv2 as cv
import matplotlib.pyplot as plt

def main():
    # Set parameters
    dispLevels = 16 #disparity range: 0 to dispLevels-1
    lambda_ = 10 #weight of smoothness cost
    trunc = 2 #truncation of smoothness cost

    # Define matching cost function
    computeMatchingCost = lambda left,right: np.absolute(left-right) #absolute differences

    # Define smoothness cost function
    computeSmoothnessCost = lambda d1,d2: lambda_*np.minimum(np.absolute(d1-d2),trunc)

    # Load left and right images in grayscale
    leftImg = cv.imread("left.png",cv.IMREAD_GRAYSCALE)
    rightImg = cv.imread("right.png",cv.IMREAD_GRAYSCALE)

    # Apply a Gaussian filter
    leftImg = cv.GaussianBlur(leftImg,(5,5),0.6)
    rightImg = cv.GaussianBlur(rightImg,(5,5),0.6)

    # Get the size
    (rows,cols) = leftImg.shape

    # Convert to int32
    leftImg = leftImg.astype(np.int32)
    rightImg = rightImg.astype(np.int32)

    # Compute pixel-based matching costs
    matchingCosts = np.zeros((rows,cols,dispLevels),dtype=np.int32)
    for d in range(dispLevels):
        rightImgShifted = shiftRight(rightImg,d,0)
        matchingCosts[:,:,d] = computeMatchingCost(leftImg,rightImgShifted)

    # Compute smoothness costs
    d = np.arange(dispLevels)
    smoothnessCosts = computeSmoothnessCost(d,d[np.newaxis,:].T)
    smoothnessCosts = smoothnessCosts[np.newaxis,np.newaxis,:,:].astype(np.int32)

    # Initialize minimum cost paths and transitions for the left to right direction
    fromLeft = np.zeros((rows,cols,dispLevels),dtype=np.int32)
    transitions = np.zeros((rows,cols,dispLevels),dtype=np.int32)

    # Compute minimum cost paths and transitions for left to right direction
    for x in range(cols-1):
        currentCosts = (matchingCosts[:,x,:] + fromLeft[:,x,:])[:,np.newaxis,:]
        C,T = computeDirectionalCosts(currentCosts,smoothnessCosts)
        fromLeft[:,x+1,:] = C[:,0,:]
        transitions[:,x+1,:] = T[:,0,:]

    # Compute the disparity map - Backtracking
    dispMap = np.zeros((rows,cols))
    ind = np.argmin(fromLeft[:,cols-1,:],axis=1)
    for x in range(cols-1,-1,-1):
        dispMap[:,x] = ind
        ind = transitions[np.arange(rows),x,ind] #get the disparity transitions

    # Normalize the disparity map for display
    scaleFactor = 256/dispLevels
    dispImg = (dispMap*scaleFactor).astype(np.uint8)

    # Show disparity map
    plt.imshow(dispImg,cmap="gray")
    plt.show(block=False)
    plt.pause(0.01)

    # Save disparity map
    cv.imwrite("disparity2_DP.png",dispImg)

    plt.show()

# Compute minimum cost paths and transitions
# ------------------------------------------
def computeDirectionalCosts(currentCosts,smoothnessCosts):
    sum_ = currentCosts[:,:,:,np.newaxis] + smoothnessCosts
    output = np.amin(sum_,axis=2)
    output = output - np.amin(output,axis=2)[:,:,np.newaxis] #normalize
    transitions = np.argmin(sum_,axis=2).astype(np.int32)
    return output,transitions

# Shift Functions (Down/Up/Right/Left/Forward/Backward)
# -----------------------------------------------------
def shiftDown(A,n,fillValue):
    B = np.roll(A,n,0)
    B[:n] = fillValue
    return B

def shiftUp(A,n,fillValue):
    B = np.roll(A,-n,0)
    B[-n:] = fillValue
    return B

def shiftRight(A,n,fillValue):
    B = np.roll(A,n,1)
    B[:,:n] = fillValue
    return B

def shiftLeft(A,n,fillValue):
    B = np.roll(A,-n,1)
    B[:,-n:] = fillValue
    return B

def shiftForward(A,n,fillValue):
    B = np.roll(A,n,2)
    B[:,:,:n] = fillValue
    return B

def shiftBackward(A,n,fillValue):
    B = np.roll(A,-n,2)
    B[:,:,-n:] = fillValue
    return B

if __name__ == "__main__":
    main()
